(()=>{"use strict";const e=new Map([[2,"AMPLIFICATION"],[1,"GAIN"],[0,"DIPLOID"],[-1,"DELETION"],[-2,"LOSS"]]),t=(e,t,n,l)=>e.filter((e=>t[e.studyId]?.[n]?.[l])).map((e=>({sampleId:e.sampleId,molecularProfileId:t[e.studyId][n][l]}))),n=e=>e.reduce(((e,t)=>t.uniqueSampleKey?(void 0===e[t.uniqueSampleKey]&&(e[t.uniqueSampleKey]=[]),e[t.uniqueSampleKey].push(t),e):e),{}),l=e=>e.map((e=>({entrezGeneId:e.entrezGeneId,hugoGeneSymbol:e.hugoGeneSymbol}))),a=(t,n,l,a,o)=>{const i=[];return i.push(n),l.forEach((n=>{let l=t[n];l&&"alteration"===n&&(l=e.get(l)),i.push(l||"")})),a.forEach((e=>{const n=o.get(e),l=n.namespace,a=n.columnName,r=((t.namespaceColumns||{})[l]||{})[a];i.push(r||"")})),i},o=e=>{try{return btoa(e)}catch(t){return Buffer.from(e).toString("base64")}},i="COPY_NUMBER_ALTERATION",r=["pyenduri@o.com","gentardis_monitoring@thehyve.nl"],s=`${studyViewPageStore.downloadFilenamePrefix}genomic_data.tsv`,c="gh_selected_columns_map",d={method:"POST",headers:{"Content-Type":"application/json"},accept:"application/json"},u=window.location.origin,m="CNA Continuous",p="CNA Continuous (log2)",g="continuousValue",y="continuousValueLog2",h=async(e,t,n)=>{if(0===t.length||0===n.length)return[];const l=!(e.includes("mutation")||e.includes("structural-variant")),a=t.reduce(((e,t)=>(void 0===e[t.molecularProfileId]&&(e[t.molecularProfileId]=[]),e[t.molecularProfileId].push(l?t.sampleId:{sampleId:t.sampleId,molecularProfileId:t.molecularProfileId}),e)),{}),o=Object.keys(a),i=await o.map((async t=>{const o=e.replace("@molecularProfileId@",t),i=l?{sampleIds:a[t]}:{sampleMolecularIdentifiers:a[t]},r={entrezGeneIds:n,...i},s=await fetch(o,{...d,body:JSON.stringify(r)});return await s.json()}));return(await Promise.all(i)).flat()},f=async(e,n,a,o,r)=>{const s=l(e),c=l(n),d=l(a),u=(e=>e.reduce(((e,t)=>t.molecularAlterationType&&t.datatype?(void 0===e[t.studyId]&&(e[t.studyId]={}),void 0===e[t.studyId][t.molecularAlterationType]&&(e[t.studyId][t.molecularAlterationType]={}),e[t.studyId][t.molecularAlterationType][t.datatype]=t.molecularProfileId,e):e),{}))(o),m=t(r,u,"MUTATION_EXTENDED","MAF"),p=t(r,u,i,"DISCRETE"),g=t(r,u,i,"CONTINUOUS"),y=t(r,u,i,"LOG2-VALUE"),h=t(r,u,"STRUCTURAL_VARIANT","SV"),[f,w,I]=await Promise.all([b(m,s),v(p,g,y,c),S(h,d)]);return{mutationData:f,cnaData:w,svData:I}},b=async(e,t)=>{if(0===e.length||0===t.length)return{};const l=t.map((e=>e.entrezGeneId)),a=await h(`${u}/api/mutations/fetch`,e,l),o=t.reduce(((e,t)=>(e[t.entrezGeneId]=t.hugoGeneSymbol,e)),{});a.forEach((e=>e.hugoGeneSymbol=o[e.entrezGeneId]));return n(a)},v=async(e,t,l,a)=>{if(0===e.length||0===a.length)return{};const o=a.map((e=>e.entrezGeneId)),i=h(`${u}/api/molecular-profiles/@molecularProfileId@/discrete-copy-number/fetch`,e,o),r=h(`${u}/api/molecular-profiles/@molecularProfileId@/molecular-data/fetch`,t,o),s=h(`${u}/api/molecular-profiles/@molecularProfileId@/molecular-data/fetch`,l,o),[c,d,m]=await Promise.all([i,r,s]),p=a.reduce(((e,t)=>(e[t.entrezGeneId]=t.hugoGeneSymbol,e)),{});c.forEach((e=>e.hugoGeneSymbol=p[e.entrezGeneId]));const f=n(c),b=n(d),v=n(m);return Object.entries(f).forEach((e=>{const t=e[0],n=e[1],l=b[t];n.forEach((e=>{const n=l?.find((t=>t.entrezGeneId===e.entrezGeneId)),a=v[t]?.find((t=>t.entrezGeneId===e.entrezGeneId));n&&(e[g]=n.value),a&&(e[y]=a.value)}))})),f},S=async(e,t)=>{if(0===e.length||0===t.length)return{};const l=t.map((e=>e.entrezGeneId)),a=await h(`${u}/api/structural-variant/fetch`,e,l),o=t.reduce(((e,t)=>(e[t.entrezGeneId]=t.hugoGeneSymbol,e)),{});a.forEach((e=>e.hugoGeneSymbol=o[e.entrezGeneId]));return n(a)},w=()=>{const e=document.getElementById("columnSelection"),t=A(e),n=JSON.stringify([...t]);localStorage.setItem(c,n)},I=(e,t,n,l)=>{const a="base-column"===n,o=(n.includes("mutation-column")||n.includes("cna-column")||n.includes("sv-column"))&&("site1HugoSymbol"===t||"site2HugoSymbol"===t),i=a||o?"disabled":"";return`<div style="display: flex; align-items: center; margin-left: 13px" class="column-selector ${n}"><input type="checkbox" id="${t}" style="margin: 4px" onclick="gh_onClickColumnCheckbox('${n}')" ${void 0===l||l.includes(t)?"checked":""} ${i}><label style="margin: 4px; font-weight: normal" for="${t}">${e}</label></div>`},C=(e,t,n=!1)=>`<div style="display: flex; align-items: center;" class="column-toggle ${t}"><input type="checkbox" style="margin: 4px" onclick="gh_toggleAll(this, '${t}')" checked ${n?"disabled":""}><label style="margin: 4px; font-weight: normal; font-style: italic ">${e}</label><span class="caret" style="margin-left: auto"/></div>`,A=e=>{const t=e.getElementsByClassName("column-selector");return["base-column","clinical-column","mutation-column","cna-column","sv-column","namespace-column"].reduce(((e,n)=>{const l=Array.prototype.slice.call(t).filter((e=>e.classList.contains(n)&&e.getElementsByTagName("input")[0].checked)).reduce(((e,t)=>{const n=t.getElementsByTagName("label")[0],l=n.textContent,a=n.getAttribute("for");return e.set(l,a),e}),new Map);return e.set(n,l),e}),new Map)},k=window.onMobxPromise,P=new Map([["Patient ID","patientId"],["Sample ID","sampleId"],["Study ID","studyId"],["Gene (HUGO)","hugoGeneSymbol"]]),G=new Map([["Center","center"],["Mutation Status","mutationStatus"],["Validation Status","validationStatus"],["Start Pos","startPosition"],["End Pos","endPosition"],["Reference Allele","referenceAllele"],["Protein Change","proteinChange"],["Mutation Type","mutationType"],["NCBI Build","ncbiBuild"],["Variant Type","variantType"],["Keyword","keyword"],["Chromosome","chr"],["Var","variantAllele"],["RefSeq mRNA ID","refseqMrnaId"],["Protein Pos Start","proteinPosStart"],["Protein Pos End","proteinPosEnd"]]),x=new Map([["CNA Discrete","alteration"],[m,g],[p,y]]),E=new Map([["Gene 1 (HUGO)","site1HugoSymbol"],["Gene 2 (HUGO)","site2HugoSymbol"],["Status","svStatus"],["Site1 Chromosome","site1Chromosome"],["Site2 Chromosome","site2Chromosome"],["Site1 Contig","site1Contig"],["Site2 Contig","site2Contig"],["Site1 Region","site1Region"],["Site2 Region","site2Region"],["Site1 Region Number","site1RegionNumber"],["Site2 Region Number","site2RegionNumber"],["Site1 Entrez Gene ID","site1EntrezGeneId"],["Site2 Entrez Gene ID","site2EntrezGeneId"],["Site1 Description","site1Description"],["Site2 Description","site2Description"],["Site1 Effect On Frame","site1EffectOnFrame"],["Site2 Effect On Frame","site2EffectOnFrame"],["Site1 Ensembl Transcript ID","site1EnsemblTranscriptId"],["Site2 Ensembl Transcript ID","site2EnsemblTranscriptId"],["Site1 Position","site1Position"],["Site2 Position","site2Position"],["Site1 Position","site1Position"],["Site2 Position","site2Position"],["Additional Annotation","annotation"],["Breakpoint Type","breakpointType"],["Connection Type","connectionType"],["Event Info","eventInfo"],["Variant Class","variantClass"],["Length","length"],["Comments","comments"],["DNA Support","dnaSupport"],["RNA Support","rnaSupport"],["Normal Read Count","normalReadCount"],["Tumor Read Count","tumorReadCount"],["Normal Variant Count","normalVariantCount"],["Tumor Variant Count","tumorVariantCount"],["Normal Paired End Read Count","normalPairedEndReadCount"],["Tumor Paired End Read Count","tumorPairedEndReadCount"],["Normal Split Read Count","normalSplitReadCount"],["Tumor Split Read Count","tumorSplitReadCount"]]);window.gh_namespaceColumnHeaderMap=new Map,window.renderCustomTab1=async e=>{N(e,!1)&&k([studyViewPageStore.mutatedGeneTableRowData,studyViewPageStore.cnaGeneTableRowData,studyViewPageStore.structuralVariantGeneTableRowData,studyViewPageStore.selectedSamples,studyViewPageStore.molecularProfiles,studyViewPageStore.clinicalAttributes],(async(t,n,l,a,o,i)=>{if(0===a.length)return;$(e).append('<div id="loadingTab" style="display: flex; flex-direction: column; align-items: center; justify-content: center; height: 100vh;">\n    <div style="color: #3786C2; width: 100px; height: 100px; margin-left: 20px; border-radius: 50%; background-color: #3786C2; display: flex; align-items: center; justify-content: center;" class="sk-spinner line-scale-pulse-out">\n        <div class="sk-rect sk-rect1" style="background-color: white; margin-right: 5px;"></div>\n        <div class="sk-rect sk-rect2" style="background-color: white; margin-right: 5px;"></div>\n        <div class="sk-rect sk-rect3" style="background-color: white; margin-right: 5px;"></div>\n        <div class="sk-rect sk-rect4" style="background-color: white; margin-right: 5px;"></div>\n        <div class="sk-rect sk-rect5" style="background-color: white;"></div>\n    </div>\n    <h4 style="margin-top: 20px;">Retrieving Data</h4>\n</div>\n');const{mutationData:r,cnaData:s,svData:c}=await f(t,n,l,o,a),d=i.reduce(((e,t)=>(e.set(t.displayName,t.clinicalAttributeId),e)),new Map);gh_namespaceColumnHeaderMap=((e,t,n)=>[...e,...t,...n].reduce(((e,t)=>(t.namespaceColumns&&Object.keys(t.namespaceColumns).length>0&&Object.entries(t.namespaceColumns).forEach((t=>{const n=t[0];Object.keys(t[1]).forEach((t=>e.set(`${n}_${t}`,{namespace:n,columnName:t})))})),e)),new Map))(Object.values(r).flat(),Object.values(s).flat(),Object.values(c).flat());const u=((e,t)=>{const n=Object.values(t).flat().some((e=>void 0!==e[g])),l=Object.values(t).flat().some((e=>void 0!==e[y])),a=new Map(e);return n||a.delete(m),l||a.delete(p),a})(x,s),h=(await T(P,d,G,u,E,gh_namespaceColumnHeaderMap),document.getElementById("loadingTab"));h&&(h.style.display="none"),D(e,P,d,G,u,E,Object.keys(gh_namespaceColumnHeaderMap))}),1)};const T=async(e,t,n,l,a,o)=>new Promise((i=>{i(null!==e&&null!==t&&null!==n&&null!==l&&null!==a&&null!==o)})),D=(e,t,n,l,a,o)=>{const i=C("All","base-column",!0),r=C("All","clinical-column"),s=C("All","mutation-column"),d=C("All","cna-column"),u=C("All","sv-column"),m=C("All","namespace-column"),p=(()=>{const e=localStorage.getItem(c);if(!e)return;return new Map(JSON.parse(e))})(),g=Array.from(t.entries()).map((e=>I(e[0],e[1],"base-column",p?.get("base-column").map((e=>e[1]))))),y=Array.from(n.entries()).map((e=>I(e[0],e[1],"clinical-column",p?.get("clinical-column").map((e=>e[1]))))),h=Array.from(l.entries()).map((e=>I(e[0],e[1],"mutation-column",p?.get("mutation-column").map((e=>e[1]))))),f=Array.from(a.entries()).map((e=>I(e[0],e[1],"cna-column",p?.get("cna-column").map((e=>e[1]))))),b=Array.from(o.entries()).map((e=>I(e[0],e[1],"sv-column",p?.get("sv-column").map((e=>e[1]))))),v=Array.from(gh_namespaceColumnHeaderMap.keys()).map((e=>I(e,e,"namespace-column",p?.get("namespace-column").map((e=>e[1]))))),S=`<div id="study-view-custom-tab">\n        <h1>Genomic data export</h1> \n        <div><p>Download genomic alterations for samples.</p></div> \n        <div style="display: flex; align-items: center">\n            <button type="button" class="btn btn-primary" style="margin-right: 5px; margin-top: 10px; width: 100px" onclick="gh_triggerDownload()">Download</button> \n            <div id="download-spinner" style="color:#3786C2;display:none" class="sk-fade-in sk-spinner sk-double-bounce"><div></div><div></div></div> \n        </div>\n        <h2>Columns</h2> \n        <div style="margin-bottom: 20px"><p>Select columns to be included in the file download.</p></div>\n        <div id="columnSelection" style="display: flex; flex-direction: row; margin-left: 15px"> \n            <div id="baseColumnsSelection" style="margin-right: 30px; min-width: 200px">\n                <h3>Base</h3>\n                <div style="width: auto; display: block;" class="btn btn-default dropdown-toggle dropdown-toggle-default">\n                    ${i}\n                </div>\n                <div style="visibility: hidden; display: flex;flex-direction: column" class="column-entries base-column">\n                    ${g.join("")}\n                </div>\n            </div>\n            <div id="ClinicalColumnsSelection" style="margin-right: 30px">\n                <h3>Clinical</h3>\n                <div style="width: auto; display: block;" class="btn btn-default dropdown-toggle dropdown-toggle-default">\n                    ${r}\n                </div>\n                <div style="visibility: hidden; flex-direction: column; width: auto; display: block;" class="column-entries clinical-column">\n                    ${y.join("")}\n                </div>\n            </div>\n            <div id="MutationColumnsSelection" style="margin-right: 30px; min-width: 200px">\n                <h3>Mutation</h3>\n                <div style="width: auto; display: block;" class="btn btn-default dropdown-toggle dropdown-toggle-default">\n                    ${s}\n                </div>\n                <div style="visibility: hidden; flex-direction: column; width: auto; display: block;" class="column-entries mutation-column">\n                    ${h.join("")}\n                </div>\n            </div> \n            <div id="CnaColumnsSelection" style="margin-right: 30px; min-width: 200px">\n                <h3>Copy Number Alteration</h3>\n                <div style="width: auto; display: block;" class="btn btn-default dropdown-toggle dropdown-toggle-default">\n                    ${d}\n                </div>\n                <div style="visibility: hidden; flex-direction: column; width: auto; display: block;" class="column-entries cna-column">\n                    ${f.join("")}\n                </div>\n            </div> \n            <div id="SvColumnsSelection" style="margin-right: 30px; min-width: 200px">\n                <h3>Structural Variant</h3>\n                <div style="width: auto; display: block;" class="btn btn-default dropdown-toggle dropdown-toggle-default">\n                    ${u}\n                </div>\n                <div style="visibility: hidden; flex-direction: column; width: auto; display: block;" class="column-entries sv-column">\n                    ${b.join("")}\n                </div>\n              </div> \n            <div id="NamespaceColumnsSelection" style="margin-right: 30px; min-width: 200px">\n                <h3>Guardant Health</h3>\n                <div style="width: auto; display: block;" class="btn btn-default dropdown-toggle dropdown-toggle-default">\n                    ${m}\n                </div>\n                <div style="visibility: hidden; flex-direction: column; width: auto; display: block;" class="column-entries namespace-column">\n                    ${v.join("")}\n                </div>\n            </div> \n        </div>`;$(e).append(S),[["baseColumnsSelection","base-column"],["ClinicalColumnsSelection","clinical-column"],["MutationColumnsSelection","mutation-column"],["CnaColumnsSelection","cna-column"],["SvColumnsSelection","sv-column"],["NamespaceColumnsSelection","namespace-column"]].forEach((([e,t])=>{R(e,t)})),["base-column","clinical-column","mutation-column","cna-column","sv-column","namespace-column"].forEach((e=>{gh_onClickColumnCheckbox(e)}))},R=(e,t)=>{document.querySelectorAll(`#${e} .btn.dropdown-toggle-default`)[0].addEventListener("click",(e=>{gh_onToggleMenu(e,t)}))};window.gh_onToggleMenu=(e,t)=>{if("INPUT"===e.srcElement.tagName)return;const n=document.querySelectorAll(`.column-entries.${t}`)[0];n.style.visibility="hidden"===n.style.visibility?"visible":"hidden"},window.gh_toggleAll=(e,t)=>{if("checkbox"!==e.type)return;const n=document.querySelectorAll(`.column-toggle.${t} input`)[0].checked;document.querySelectorAll(`.column-selector.${t} input`).forEach((e=>{e.disabled||(e.checked=n)})),w()},window.gh_onClickColumnCheckbox=e=>{const t=Array.prototype.slice.call(document.querySelectorAll(`.column-selector.${e} input`)).filter((e=>!1===e.disabled)),n=Array.prototype.slice.call(t).filter((e=>e.checked&&!1===e.disabled)),l=n.length===t.length,a=0===n.length,o=document.querySelectorAll(`.column-toggle.${e} input`)[0];l||a?l?(o.checked=!0,o.indeterminate=!1):(o.checked=!1,o.indeterminate=!1):(o.checked=!1,o.indeterminate=!0),w()},window.gh_triggerDownload=async()=>{const e=document.getElementById("download-spinner");e.style.display="block",k([studyViewPageStore.mutatedGeneTableRowData,studyViewPageStore.cnaGeneTableRowData,studyViewPageStore.structuralVariantGeneTableRowData,studyViewPageStore.selectedSamples,studyViewPageStore.molecularProfiles,studyViewPageStore.getDataForClinicalDataTab],(async(t,n,l,i,r,c)=>{if(0===i.length)return;const{mutationData:d,cnaData:u,svData:m}=await f(t,n,l,r,i),p=((e,t,n,l,i,r,s)=>{if(0===e?.length)return;const c=((e,t)=>{const n=Array.from(t.get("base-column").values()).filter((e=>"hugoGeneSymbol"!==e)),l=Array.from(t.get("clinical-column").values());return e.reduce(((e,t)=>(e[(e=>{if(!e.sampleId||!e.studyId)throw Error("Object must have sampleId and studyId");return o(`${e.sampleId}:${e.studyId}`).replace(/={1,2}$/,"")})(t)]=[...n,...l].map((e=>t[e]||"")),e)),{})})(r,s),d=Array.from(s.get("clinical-column").values()),u=Array.from(s.get("mutation-column").values()),m=Array.from(s.get("cna-column").values()),p=Array.from(s.get("sv-column").values()),g=Array.from(s.get("namespace-column").values()),y=["hugoGeneSymbol",...u,...m,...p],h=e.reduce(((e,o)=>{void 0===e[o.uniqueSampleKey]&&(e[o.uniqueSampleKey]={});const r=i.reduce(((e,i)=>{let r=t[o.uniqueSampleKey]?.filter((e=>e.entrezGeneId===i.entrezGeneId)),s=n[o.uniqueSampleKey]?.filter((e=>e.entrezGeneId===i.entrezGeneId)),c=l[o.uniqueSampleKey]?.filter((e=>e.site1EntrezGeneId===i.entrezGeneId||e.site2EntrezGeneId===i.entrezGeneId));const d=u.length>0&&r?.length>0,h=m.length>0&&s?.length>0,f=p.length>2&&c?.length>0;return d&&e.push(...r.map((e=>a(e,"MUT",y,g,gh_namespaceColumnHeaderMap)))),h&&e.push(...s.map((e=>a(e,"CNA",y,g,gh_namespaceColumnHeaderMap)))),f&&e.push(...c.map((e=>a(e,"SV",y,g,gh_namespaceColumnHeaderMap)))),e}),[]);return e[o.uniqueSampleKey]=r,e}),{}),f=Array.from(s.get("base-column").entries()).filter((e=>"hugoGeneSymbol"!==e[1])).map((e=>e[0])),b=Array.from(s.get("clinical-column").keys()),v=Array.from(s.get("mutation-column").keys()),S=Array.from(s.get("cna-column").keys());let w=Array.from(s.get("sv-column").keys());w=2===w.length?[]:w;const I=Array.from(s.get("namespace-column").keys()),C=2+v.length+S.length+w.length+I.length,A=Array(C).fill(""),k=[[...f,...b,"Alteration Type","Gene",...v,...S,...w,...I]];return e.forEach((e=>{let t=h[e.uniqueSampleKey],n=c[e.uniqueSampleKey];const l=d.length;(n||t.length>0)&&(t=t?.length>0?t:[A],n=n?.length>0?n:[e.patientId,e.sampleId,e.studyId,...Array(l).fill("")],t.forEach((e=>{k.push([...n,...e])})))})),k})(i,d,u,m,[...t,...n,...l].map((e=>({entrezGeneId:e.entrezGeneId,hugoGeneSymbol:e.hugoGeneSymbol}))),c,A(document.getElementById("columnSelection"))),g=p.map((e=>e.join("\t"))).join("\n");e.style.display="none";var y=document.createElement("a");y.setAttribute("href","data:text/plain;charset=utf-8, "+encodeURIComponent(g)),y.setAttribute("download",s),y.setAttribute("target","_blank"),document.body.appendChild(y),y.click(),document.body.removeChild(y)}),1)};const N=(e,t=!1)=>{const n=["mutatedGeneTableRowData","cnaGeneTableRowData","structuralVariantGeneTableRowData","selectedSamples","molecularProfiles","clinicalAttributes","getDataForClinicalDataTab"].filter((e=>void 0===studyViewPageStore[e]));return!(n.length>0||t)||(t&&(n.push("placeholder1"),n.push("placeholder2")),$(e).append(`<div class="alert alert-danger" style="padding: 50px; display: flex">\n        <div>\n            <i class="fa fa-1x fa-times" style="margin-right: 6px; margin-bottom: 1px;"></i>\n        </div>\n        <div style="padding-left: 5px">\n            Oops, something went wrong! Required methods are missing (or were renamed) from the StudyViewPageStore:\n            <ul style="margin-top: 5px; margin-bottom: 5px">\n            <li>${n.join("</li><li>")}</li>\n            </ul>\n            Please report this error <a href="mailto:${r.join(",")}">here</a>.\n        </div>\n       </div>`),!1)}})();
